lib/puppet/provider/iam_instance_profile/v2.rb

```diff
@@ -0,0 +1,93 @@
+require_relative '../../../puppet_x/puppetlabs/aws.rb'
+
+require 'uri'
+
+Puppet::Type.type(:iam_instance_profile).provide(:v2, :parent => PuppetX::Puppetlabs::Aws) do
+  confine feature: :aws
+
+  mk_resource_methods
+
+  read_only(:arn)
+
+  def self.instances
+    response = iam_client.list_instance_profiles()
+    response.instance_profiles.collect do |instance_profile|
+      new({
+              name: instance_profile.instance_profile_name,
+              ensure: :present,
+              path: instance_profile.path,
+              arn: instance_profile.arn,
+              roles: instance_profile.roles,
```

Comment:
On puppet 3 the type with `:array_matching => :all` will return a single non-array value whether it is passed a single value or an array of a single value. On puppet 4 the type will always return an array of single or multiple values.

The problem is that if your provider `self.instances` returns roles as an array of a single item, puppet 3 will try to turn `['value']` into `'value'` and cause changes every run. You can document this property as only working on puppet 4. Making it work on 3 is just icky.